kaitai/kaitaistream.cpp

```diff
@@ -691,9 +694,143 @@ std::string kaitai::kstream::bytes_to_str(std::string src, std::string src_enc)
+    std::string enc(src_enc);
+    if (enc == "UTF-8") {
+        return CP_UTF8;
+    } else if (enc == "UTF-16LE") {
+        return KAITAI_CP_UTF16LE;
+    } else if (enc == "UTF-16BE") {
+        return KAITAI_CP_UTF16BE;
+    } else if (enc == "IBM437") {
+        return 437;
+    } else if (enc == "IBM850") {
+        return 850;
+    } else if (enc == "SHIFT_JIS") {
+        return 932;
+    } else if (enc == "GB2312") {
+        return 936;
+    } else if (enc == "ASCII") {
+        return 20127;
+    } else if (enc == "EUC-JP") {
+        return 20932;
+    } else if (enc == "ISO-8859-1") {
+        return 28591;
+    } else if (enc == "ISO-8859-2") {
+        return 28592;
+    } else if (enc == "ISO-8859-3") {
+        return 28593;
+    } else if (enc == "ISO-8859-4") {
+        return 28594;
+    } else if (enc == "ISO-8859-5") {
+        return 28595;
+    } else if (enc == "ISO-8859-6") {
+        return 28596;
+    } else if (enc == "ISO-8859-7") {
+        return 28597;
+    } else if (enc == "ISO-8859-8") {
+        return 28598;
+    } else if (enc == "ISO-8859-9") {
+        return 28599;
+    } else if (enc == "ISO-8859-10") {
+        return 28600;
+    } else if (enc == "ISO-8859-11") {
+        return 28601;
+    } else if (enc == "ISO-8859-13") {
+        return 28603;
+    } else if (enc == "ISO-8859-14") {
+        return 28604;
+    } else if (enc == "ISO-8859-15") {
+        return 28605;
+    } else if (enc == "ISO-8859-16") {
+        return 28606;
+    }
+
+    return KAITAI_CP_UNSUPPORTED;
+}
+
+std::string kaitai::kstream::bytes_to_str(const std::string src, const char* src_enc) {
+    // Step 1: convert encoding name to codepage number
+    int codepage = encoding_to_win_codepage(src_enc);
+    return bytes_to_str(src, codepage);
+}
+
+std::string kaitai::kstream::bytes_to_str(const std::string src, int codepage) {
+    if (codepage == KAITAI_CP_UNSUPPORTED) {
+        throw std::runtime_error("bytes_to_str: unsupported encoding name");
+    }
+
+    // Shortcut: if we're already in UTF-8, no need to convert anything
+    if (codepage == CP_UTF8) {
+        return src;
+    }
+
+    // Step 2: convert bytes to UTF16 ("wide char") string
+    std::wstring utf16;
+    int32_t utf16_len;
+    int32_t src_len;
+    if (src.length() > std::numeric_limits<int32_t>::max()) {
+        throw std::runtime_error("bytes_to_str: buffers longer than int32_t are unsupported");
+    } else {
+        src_len = static_cast<int32_t>(src.length());
+    }
+
+    switch (codepage) {
+    case KAITAI_CP_UTF16LE:
+        // If our source is already UTF16LE, just copy it
+        utf16_len = src_len / 2;
+        utf16 = std::wstring((wchar_t*)src.c_str(), utf16_len);
+        break;
+    case KAITAI_CP_UTF16BE:
+        // If our source is in UTF16BE, convert it to UTF16LE by swapping bytes
+        utf16_len = src_len / 2;
+
+        utf16 = std::wstring(utf16_len, L'\0');
+        for (int i = 0; i < utf16_len; i++) {
```

Comment:
In practice, `int` is perhaps always the same as `int32_t` _on Windows_, so it won't make much of a difference, but I suppose `i` should be rather `size_t` or at least `int32_t` (which is apparently enough in this case because `src_len` is at most `std::numeric_limits<int32_t>::max()` and `utf16_len = src_len / 2`).

Reply:
Good catch! Let me at least make it int32_t everywhere.